lighthouse-core/gather/gatherers/image-elements.js

```diff
@@ -308,30 +312,39 @@ class ImageElements extends Gatherer {
 
     /** @type {Array<LH.Artifacts.ImageElement>} */
     const imageUsage = [];
-    const top50Images = Object.values(indexedNetworkRecords)
-      .sort((a, b) => b.resourceSize - a.resourceSize)
-      .slice(0, 50);
+
     await Promise.all([
       driver.sendCommand('DOM.enable'),
       driver.sendCommand('CSS.enable'),
       driver.sendCommand('DOM.getDocument', {depth: -1, pierce: true}),
     ]);
 
+    // Sort (in-place) as largest images descending
+    elements.sort((a, b) => {
+      const aRecord = indexedNetworkRecords[a.src] || {};
+      const bRecord = indexedNetworkRecords[b.src] || {};
+      return bRecord.resourceSize - aRecord.resourceSize;
+    });
+
+    // Don't allow more than 3s of this expensive protocol work
+    let hasBudget = true;
```

Comment:
total bikeshed: `metBudget`? `metProtocolBudget`? `reachedGatheringBudget`?

`hasBudget` in the conditionals below sounds like some aspect of the element being examined rather than part of the gatherer control flow

Reply:
good call.  it was "hasRemainingBudget" which is definitely way more clear. 

i like your inverted names. i'll noodle.

Reply:
`hasRemainingCDPBudget` ?  i prefer checking the optimistic name in the conditionals

Reply:
it'd be lovely if a time budget were built in to all gatherers...

Would require a yielding syntax I think..

Reply:
> it'd be lovely if a time budget were built in to all gatherers...

yah i actually wrote it first like that. But then abandoned because it was like 15 lines of moderate complexity and this new one are these really boring 2 lines. 

Also these image-elem cases are interesting because it's just these two methods (`fetchSourceRules` and `fetchElementWithSizeInformation`) that we want the budget to apply to. everything before that in the gatherer is excluded from the budget.  (basically, the budget doesn't apply to the entire gatherer.) 

i was thinking about a budget for running axe actually. so when i explore that i'll try to extract a common component.

Reply:
> `hasRemainingCDPBudget` ? i prefer checking the optimistic name in the conditionals

yeah, I like the optimistic take :) Any name like that works for me.

Reply:
> i was thinking about a budget for running axe actually. so when i explore that i'll try to extract a common component.

this frightens me (arbitrary gatherer teardown?), but maybe with FR and everyone gets a `Session` or something it's possible :)

Reply:
That was exactly my thought. We give everyone a fresh session object with a timer that starts throwing after X seconds :) 

Doesn't help the axe case though where our shared execution context is still pegged on axe